gnrc/tcp : Expose configurations to Kconfig

akshaim opened the pull request:
### Contribution description

This PR exposes compile configurations in GNRC: TCP to Kconfig.

### Testing procedure

1. New documentation was built using Doxygen 

The build works fine.

2. Header and macro was introduced to tests/gnrc_tcp/

```
#include "net/gnrc/tcp/config.h"
#define STR(x)   #x
#define SHOW_DEFINE(x) printf("%s=%s\n", #x, STR(x))

```
The test file can be found [here](https://github.com/akshaim/RIOT/blob/82bdac7d599dd692f3e090a21d9b8468d6419c78/tests/gnrc_tcp/main.c)

Compiled binaries (BOARD=samr21-xpro) were uploaded to FIT IoT Test bed.

#### Default State:

##### Firmware Output

~~}!main(): This is RIOT! (Version: 2020.07-devel-737-g0b75d-Kconfig_tcp_tests)
~~}!CONFIG_GNRC_TCP_CONNECTION_TIMEOUT_DURATION=3000000
~~}!CONFIG_GNRC_TCP_MSL=1000000
~~}!CONFIG_GNRC_TCP_MSS_MULTIPLICATOR=(1U)
~~}!CONFIG_GNRC_TCP_RCV_BUFFERS=(1U)
~~}!CONFIG_GNRC_TCP_RTO_LOWER_BOUND=(1U * (1000000LU))
~~}!CONFIG_GNRC_TCP_RTO_UPPER_BOUND=(60U * (1000000LU))
~~}!CONFIG_GNRC_TCP_RTO_GRANULARITY=(10U * (1000LU))
~~}!CONFIG_GNRC_TCP_RTO_A_DIV=(8U)
~~}!CONFIG_GNRC_TCP_RTO_B_DIV=(4U)
~~}!CONFIG_GNRC_TCP_RTO_K=(4U)
~~}!CONFIG_GNRC_TCP_PROBE_LOWER_BOUND=(1U * (1000000LU))
~~}!CONFIG_GNRC_TCP_PROBE_UPPER_BOUND=(60U * (1000000LU))
~~}!RIOT GNRC_TCP test application

#### Usage with menuconfig [default values]

> make menuconfig

##### Firmware Output

!main(): This is RIOT! (Version: 2020.07-devel-737-g82bda-Kconfig_tcp_tests)
~~}!CONFIG_GNRC_TCP_CONNECTION_TIMEOUT_DURATION=120000000
~~}!CONFIG_GNRC_TCP_MSL=30000000
~~}!CONFIG_GNRC_TCP_MSS_MULTIPLICATOR=1
~~}!CONFIG_GNRC_TCP_RCV_BUFFERS=1
~~}!CONFIG_GNRC_TCP_RTO_LOWER_BOUND=1000000
~~}!CONFIG_GNRC_TCP_RTO_UPPER_BOUND=60000000
~~}!CONFIG_GNRC_TCP_RTO_GRANULARITY=10000
~~}!CONFIG_GNRC_TCP_RTO_A_DIV=8
~~}!CONFIG_GNRC_TCP_RTO_B_DIV=4
~~}!CONFIG_GNRC_TCP_RTO_K=4
~~}!CONFIG_GNRC_TCP_PROBE_LOWER_BOUND=1000000
~~}!CONFIG_GNRC_TCP_PROBE_UPPER_BOUND=60000000
~~}!RIOT GNRC_TCP test application

#### Usage with menuconfig

> make menuconfig

##### Firmware Output

!main(): This is RIOT! (Version: 2020.07-devel-737-g82bda-Kconfig_tcp_tests)
~~}!CONFIG_GNRC_TCP_CONNECTION_TIMEOUT_DURATION=110000000
~~}!CONFIG_GNRC_TCP_MSL=32000000
~~}!CONFIG_GNRC_TCP_MSS_MULTIPLICATOR=2
~~}!CONFIG_GNRC_TCP_RCV_BUFFERS=2
~~}!CONFIG_GNRC_TCP_RTO_LOWER_BOUND=1200000
~~}!CONFIG_GNRC_TCP_RTO_UPPER_BOUND=62000000
~~}!CONFIG_GNRC_TCP_RTO_GRANULARITY=20000
~~}!CONFIG_GNRC_TCP_RTO_A_DIV=4
~~}!CONFIG_GNRC_TCP_RTO_B_DIV=2
~~}!CONFIG_GNRC_TCP_RTO_K=2
~~}!CONFIG_GNRC_TCP_PROBE_LOWER_BOUND=1200000
~~}!CONFIG_GNRC_TCP_PROBE_UPPER_BOUND=50000000
~~}!RIOT GNRC_TCP test application


**MACROS were successfully configured.**

### Issues/PRs references

#12888

leandrolanzieri commented on sys/include/net/gnrc/tcp/config.h:
```diff
@@ -4,6 +4,8 @@
  * This file is subject to the terms and conditions of the GNU Lesser
  * General Public License v2.1. See the file LICENSE in the top level
  * directory for more details.
+ *
+ *
```
?

akshaim replied:
OOps

leandrolanzieri commented on sys/include/net/gnrc/tcp/config.h:
```diff
@@ -26,6 +28,28 @@
+ *     RTTVAR <- (1 - beta) * RTTVAR + beta * |SRTT - R'|
+ *     SRTT <- (1 - alpha) * SRTT + alpha * R'
+ *
+ * where alpha and beta are constants, and R' is the instantaneous RTT value.
+ *
+ * RTO is then calculated as :
+ *
+ *     RTO <- SRTT + max (G, K*RTTVAR)
+ *
+ * where K is a constant, and G is clock granularity in seconds.
```
Maybe you can add references to the macros that represent these parameters

akshaim replied:
Done.

leandrolanzieri commented on sys/include/net/gnrc/tcp/config.h:
```diff
@@ -102,6 +130,7 @@ extern "C" {
 
 /**
  * @brief Alpha value for RTO calculation, default is 1/8
+ *
```
Extra line

leandrolanzieri commented on sys/net/gnrc/transport_layer/tcp/gnrc_tcp.c:
```diff
@@ -186,8 +186,9 @@ static int _gnrc_tcp_open(gnrc_tcp_tcb_t *tcb, const gnrc_tcp_ep_t *remote,
         /* Setup connection timeout: Put timeout message in TCBs mbox on expiration */
-        _setup_timeout(&connection_timeout, GNRC_TCP_CONNECTION_TIMEOUT_DURATION,
-                       _cb_mbox_put_msg, &connection_timeout_arg);
+        _setup_timeout(&connection_timeout, \
```
Here and bellow, please keep the current style of the file for multiline function call. I think you can drop the `\` as well.

akshaim replied:
Okay

akshaim replied:
Done.

brummer-simon commented on sys/include/net/gnrc/tcp/config.h:
```diff
@@ -26,18 +26,43 @@
+ * where alpha ( @ref CONFIG_GNRC_TCP_RTO_A_DIV ) and beta
+ * ( @ref CONFIG_GNRC_TCP_RTO_B_DIV) are constants, and R' is the instantaneous
+ * RTT value.
+ *
```
The comment above is a little bit misleading. CONFIG_GNRC_TCP_RTO_A_DIV is not the alpha value, it is the denominator for alpha.

The default value for Alpha is 1/8, therefore  CONFIG_GNRC_TCP_RTO_A_DIV is 8. The same goes for beta and CONFIG_GNRC_TCP_RTO_B_DIV.

brummer-simon commented on sys/include/net/gnrc/tcp/config.h:
```diff
@@ -26,18 +26,43 @@
+ * where alpha ( @ref CONFIG_GNRC_TCP_RTO_A_DIV ) and beta
+ * ( @ref CONFIG_GNRC_TCP_RTO_B_DIV) are constants, and R' is the instantaneous
+ * RTT value.
+ *
```
The comment above is a little bit misleading. CONFIG_GNRC_TCP_RTO_A_DIV is not the alpha value, it is the denominator for alpha.

The default value for Alpha is 1/8, therefore CONFIG_GNRC_TCP_RTO_A_DIV is 8. The same goes for beta and CONFIG_GNRC_TCP_RTO_B_DIV.

akshaim replied:
Appreciate the comments. Updated.

brummer-simon commented on sys/include/net/gnrc/tcp/config.h:
```diff
@@ -54,22 +79,22 @@ extern "C" {
-#define GNRC_TCP_DEFAULT_WINDOW (GNRC_TCP_MSS * GNRC_TCP_MSS_MULTIPLICATOR)
+#define GNRC_TCP_DEFAULT_WINDOW (GNRC_TCP_MSS * CONFIG_GNRC_TCP_MSS_MULTIPLICATOR)
 #endif
 
```
Why is the default receive window size not configurable?

akshaim replied:
The value is now found from two configurations. Is there a need to allow the macro to be configured directly?

brummer-simon replied:
This value marks the maximum amount of bytes that can be received from the peer at a given moment.
In general it makes sence to elide this value from "n times maximum segment size" however this parameter influences TCPs behavior significantly. Making in configurable might have its benefits.

brummer-simon commented on sys/include/net/gnrc/tcp/config.h:
```diff
@@ -54,22 +79,22 @@ extern "C" {
 /**
```
Why is the maximum segment size not configurable? The MSS is usually aligned with the underlying network layer for performance reasons, however is could be customized.

akshaim replied:
I was doubtful about whether MSS can be configured. Thank you for the comment.

brummer-simon commented on sys/net/gnrc/transport_layer/tcp/Kconfig:
```diff
@@ -0,0 +1,85 @@
+        receive buffer.
+
+config GNRC_TCP_RCV_BUFFERS
+    int "Number of preallocated receive buffers"
```
The description could be better. This value determines how many parallel TCP connection can be active at the same time.

akshaim replied:
Added. Thanks for the comment

brummer-simon replied:
Its still the same description

brummer-simon commented on sys/net/gnrc/transport_layer/tcp/Kconfig:
```diff
@@ -0,0 +1,85 @@
+    default 10000
+    help
+        Clock granularity for retransmission timeout (RTO) for TCP in
+        microseconds. Default value is 10000000 microseconds (10 seconds). 
```
This is not true. The default granularity is 10 miliseconds.

akshaim replied:
Updated. Thanks

brummer-simon commented on sys/net/gnrc/transport_layer/tcp/Kconfig:
```diff
@@ -0,0 +1,85 @@
+        Refer to RFC 6298 for more information.
+
+config GNRC_TCP_RTO_A_DIV
+    int "Alpha value for RTO calculation"
```
As mentioned above it is not the alpha value. Alpha is 1 / GNRC_TCP_RTO_A_DIV

akshaim replied:
Done. Thanks.

brummer-simon commented on sys/net/gnrc/transport_layer/tcp/Kconfig:
```diff
@@ -0,0 +1,85 @@
+    int "Alpha value for RTO calculation"
+    default 8
+
+config GNRC_TCP_RTO_B_DIV
```
As mentioned above it is not the beta value. Beta is 1 / GNRC_TCP_RTO_B_DIV

akshaim replied:
Done. Thanks

akshaim commented:
> I have some small change requests aside from that good work.

Thank you for the comments/ I have updated the file. Introduced a new symbol to make sure that the default configuration remains intact. Directly configuring the `CONFIG_GNRC_TCP_DEFAULT_WINDOW` via menuconfig does not have a dependency on `GNRC_TCP_MSS_MULTIPLICATOR`.

akshaim commented:
> I have tested this PR on "native" and on "nucleo-f401re". The test suite works as expected an all configured values are present in the resulting binary. From my point of view, this PR can be merged, although I don't have the rights to do that.
> 
> @akshaim - Thanks for adding the Kconfig support.

Thanks @brummer-simon  for the review. @leandrolanzieri  Should I squash now ?

leandrolanzieri commented:
> > I have tested this PR on "native" and on "nucleo-f401re". The test suite works as expected an all configured values are present in the resulting binary. From my point of view, this PR can be merged, although I don't have the rights to do that.
> > @akshaim - Thanks for adding the Kconfig support.
> 
> Thanks @brummer-simon for the review. @leandrolanzieri Should I squash now ?

Yes please

akshaim commented:
> > > I have tested this PR on "native" and on "nucleo-f401re". The test suite works as expected an all configured values are present in the resulting binary. From my point of view, this PR can be merged, although I don't have the rights to do that.
> > > @akshaim - Thanks for adding the Kconfig support.
> > 
> > 
> > Thanks @brummer-simon for the review. @leandrolanzieri Should I squash now ?
> 
> Yes please

Done.  I have reordered the commit history, to make better sense of, but github displays commits by order of author date.

akshaim commented:
Thanks @leandrolanzieri
